dependent_types.md

```diff
@@ -0,0 +1,279 @@
+
+```
+wasItStringOrInt :: IntOrString a -> String
+wasItStringOrInt x =
+    case x of
+        -- Local assumptions - Local because it is limited to only one case branch 
+        -- and assumption because we assume what type is our type parameter
+        IntConstructor a    -> "Constructed with Int"      -- a ~ Int
+        StringConstructor b -> "Constructed with String    -- a ~ String
+        
+-- ghci
+λ> let x = IntConstructor 42
+λ> wasItStringOrInt x
+"Constructed with Int"
+```
+
+## Generic
+
+Lets now build our intuition about types in general and also dependent types.
+
+If we look at a normal Haskell Algebraic Data Type `T`
+
+```
+data T = A Int Int
+  | B String String String
+  | C () Void
+  | D
+```
+
+We can say that `T` is a sum of products `A B C D`. In order to understand sum we can define this and every other ADT only using
+`Either`, `() (unit)` , `(,) tuple`  and function type. So our ADT could be encoded like this using infix notation
+
+```
+type T' = (Int, Int) `Either` (String, String, String) `Either` ((), Void) `Either` ()
+```
+
+Tuples provide us with products, Either provides us with sums, unit is empty constructor and function type is just a built in function. 
+
+If we understand these simple types than we understand Algebraic Data Types.
+
+If we understand Σ (sigma) and Π (pi) we understand dependent types. 
+
+You can see the theory behing all of this [here](https://en.wikipedia.org/wiki/Intuitionistic_type_theory) .
+
+## Σ and Π
+
+First we will look at some pseudo code so we can explain easier what is going on and after that we will take a look at the example that actually compiles.
+
+Sigma type is like a tuple but with some caviats
+
+```
+     (A     , B) 
+Σ :: (x :: A) B(x)
+    
+```
+Here `x` is of type `A` and we are able to use it inside some type level function `B(x)`.
+
+One possible sigma type could be
+
+```
+Σ :: (x :: Bool) (if x then Int else String) -- this if statement can be implemented as a type family in haskell
+
+-- possible values
+(True, 42)
+(False, "abc")
+``` 
+And if we try to use this in a function we can have something like:
+
+```
+f :: ∑ (x :: Bool) (if x then Int else String) -> String
+f (x,y) = ???
+```
+So now we have a function from `sigma` to `String` , we know that `x` is of type `Bool` but we have no idea what `y` is .
+What can we do with it ?
+We can pattern match on it!
+
+```
+f :: (x :: Bool) (if x then Int else String) -> String
+f (x,y) = case x of
+  True -> show y -- if x is True then y ~ Int
+  False -> y     -- if x is False then y ~ String
+```
+As soon as we pattern match we are able to use local assumptions about `y` which is a term level data and we know that `y` can be either `Int` or `String`. Any other type would have caused compile time error and in this way we are able to prevent our programs from working with some random types that we don't want and prove that our program is correct at compile time!
```

Comment:
> a term level data

Same issue, "data" is plural.

Reply:
> `y` can be either `Int` or `String`

This phrasing is a bit unclear because it's the type of `y` which is either `Int` or `String`, not `y` itself.

Reply:
>  Any other type would have caused compile time error and in this way we are able to prevent our programs from working with some random types that we don't want and prove that our program is correct at compile time!

It's not like without dependent types we have random types floating around :smile: